### PR TITLE
Feat basic accounting and user activation

### PR DESCRIPTION
**Pull request by clemensmanert**
Adds basic accounting features to the backend. Existing pendng accounts can be activated.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,6 +39,17 @@ project root
 └── app.py        # The app-main.
 ```
 
+## Database
+The database structure is created using migrations. A migration is a
```
Comment (missgreenwood) [inline README.md:45]: As we discussed: Please provide information on required environment variables.

File: app.py [hunk 2]
```diff
@@ -1,15 +1,20 @@
 from os import environ
+
 from setup_app import setup_app
 
 
 class RunConfig():
     """Grabs app parameters from the environment."""
     SECRET_KEY = environ.get('BUZZN_SECRET_KEY')
     CLIENT_NAME = 'BuzznClient'
+    SQLALCHEMY_TRACK_MODIFICATIONS = False
```
Comment (missgreenwood) [inline app.py:10]: Likewise, please provide this information in the README.

File: tests/set_password_test.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+from models.user import User
+from routes.set_password import Errors
+from util.database import db
+
+
+class SetPasswordTest(BuzznTestCase):
+    """Checks if users can be created."""
+
+    def test_user_does_not_exist(self):
+        """Check whether a non existing user result in an error"""
```
Comment (missgreenwood) [inline tests/set_password_test.py]: "Check whether a non existing user results [...]". Please give information in the comment what exactly you are testing (i.e. if trying to set the password for a non-existing user results in an error).

File: tests/set_password_test.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            'user': 'SomeOtherUser',
+            'token': 'Wrong token',
+            'password': 'SomePassword'
+        }))
+        self.assertEqual(response.status_code,
+                         status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.json, Errors.WRONG_TOKEN.__dict__)
+
+    def test_no_activation_pending(self):
+        """Check whether an account can not be activated twice"""
```
Comment (missgreenwood) [inline tests/set_password_test.py]: Suggestions for the comment: 
"Check whether an account can be activated twice" or "Make sure an account cannot be activated twice" to make the intent clearer.

File: tests/set_password_test.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            'user': 'SomeUser',
+            'token': 'SomeToken',
+            'password': 'SomePassword'
+        }))
+        self.assertEqual(response.status_code,
+                         status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.json, Errors.NO_ACTIVATION_PENDING.__dict__)
+
+    def test_correct_data_should_set_password(self):
+        """Check whether correct provided parameters do activate an account"""
```
Comment (missgreenwood) [inline tests/set_password_test.py]: "[...] correctly [...]"

File: tests/buzzn_test_case.py [hunk 6]
```diff
@@ -1,17 +1,24 @@
...
     LIVESERVER_PORT = 0
     CLIENT_NAME = 'BuzznClient'
+    SQLALCHEMY_TRACK_MODIFICATIONS = False
+    SQLALCHEMY_DATABASE_URI = 'sqlite:///:memory:'
+    PASSWORD_SALT = bcrypt.gensalt()
 
 
 class BuzznTestCase(TestCase):
-    """Creates app configured to run buzzn tests."""
+    """Creates apps configures to run buzzn tests."""
```
Comment (missgreenwood) [inline tests/buzzn_test_case.py]: :) 
You mean: one app to run tests, right? I'd suggest "Creates app configured to run buzzn tests."

File: requirements.txt [hunk 7]
```diff
@@ -1,6 +1,10 @@
+bcrypt
 flask
 requests
 requests_oauthlib
 flask_api
-flask_testing
+flask_jwt_extended
+flask_migrate
+flask_sqlalchemy
+flask-testing
```
Comment (missgreenwood) [inline requirements.txt]: flask-testing or flask_testing?

File: migrations/env.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
+from __future__ import with_statement
```
Comment (missgreenwood) [inline migrations/env.py]: This is normally used for older Python versions to import something from future versions. Are you sure we need this, considering we want to use the latest Python version?

File: migrations/versions/b4218a854a1f_.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
+"""empty message
```
Comment (missgreenwood) [inline migrations/versions/b4218a854a1f_.py]: Can the "empty message" be avoided?

File: migrations/versions/b4218a854a1f_.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+# revision identifiers, used by Alembic.
+revision = 'b4218a854a1f'
+down_revision = None
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (missgreenwood) [inline migrations/versions/b4218a854a1f_.py]: Please adjust the comment to the actual functionalities.

File: migrations/versions/b4218a854a1f_.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    sa.Column('_role', sa.Enum('LOCAL_POWER_TAKER', 'ADMINISTRATOR', name='roletype'), nullable=True),
+    sa.PrimaryKeyConstraint('_id'),
+    sa.UniqueConstraint('_activation_token'),
+    sa.UniqueConstraint('_name')
+    )
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (missgreenwood) [inline migrations/versions/b4218a854a1f_.py]: Likewise, please adjust the standard comment to the actual functionalities.

File: models/user.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
+from enum import Enum
+import bcrypt
+from flask import current_app
+from util.database import db
+
+
+class RoleType(Enum):
+    """Indicates the role of a user account. This may privilige the user for
```
Comment (missgreenwood) [inline models/user.py]: "privilege"

File: models/user.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+class ActiveType(Enum):
+    """Indicates the user account's state in its lifecycle.
+    """
+    ACTIVATION_PENDING = 'User activation pending.'
+    ACTIVE = 'User account is active.'
+    DISABLED = 'User account is disabled'
+
+
+class User(db.Model):
+    """Represents an user account in the backend.
```
Comment (missgreenwood) [inline models/user.py]: "a user"

File: models/user.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+    Arguments:
```
Comment (missgreenwood) [inline models/user.py]: Please change to docstring comment style as discussed.

File: models/user.py [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+        db {class} -- Db related stuff to crud an instance of this class.
```
Comment (missgreenwood) [inline models/user.py]: to create?

File: models/user.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+        return 50
+
+    @staticmethod
+    def PASSWORD_MAX_LENGTH():
+        return 129
+
+    @staticmethod
+    def generate_password_hash(target):
+        """Generates a password hash for the given password using app's
+        spassword salt.
```
Comment (missgreenwood) [inline models/user.py:37]: I'd suggest: "[...] using the app's password salt."

File: models/user.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+
+        Arguments:
```
Comment (missgreenwood) [inline models/user.py]: Same here, please change to docstring comment format.

File: models/user.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+    _id = db.Column(db.Integer, primary_key=True)
+    _name = db.Column(db.String(33), unique=True)
+    _activation_token = db.Column(db.String(33), unique=True)
+    _password = db.Column(db.String(333))
+    _status = db.Column(db.Enum(ActiveType))
+    _role = db.Column(db.Enum(RoleType))
+
+    def __init__(self, name, activation_token):
+        """Creates a new useraccount and sets its state into pending.
```
Comment (missgreenwood) [inline models/user.py]: I'd suggest: "Creates a new user account and sets its state to pending."

File: models/user.py [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+
+        Arguments:
```
Comment (missgreenwood) [inline models/user.py]: docstring comment format

File: models/user.py [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+    def get_status(self):
+        return self._status
+
+    def get_role(self):
+        return self._role
+
+    def is_active(self):
+        """Returns a value indicating whether this account is active.
+
+        Returns:
```
Comment (missgreenwood) [inline models/user.py]: docstring comment format

File: models/user.py [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+            Boolean -- True, if this account is active, false otherwise.
```
Comment (missgreenwood) [inline models/user.py]: Just one comment on Python types: The string type is called "str", Boolean values are "True" and "False" (capital letters, also key words).

File: models/user.py [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+        """
+        return self._status == ActiveType.ACTIVE
+
+    def check_password(self, password_to_check):
+        """Checks whether the given password matches the user's.
+
+        Arguments:
```
Comment (missgreenwood) [inline models/user.py]: docstring comment style

File: models/user.py [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+            password_to_check {String} -- Password in plaintext.
+
+        Returns:
+            Boolean -- True if the passwort matches, false otherwise.
```
Comment (missgreenwood) [inline models/user.py]: Cf. my comment on Boolean values.

File: models/user.py [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+        """
+        return bcrypt.checkpw(User.generate_password_hash(password_to_check),
+                              self._password)
+
+    def set_password(self, newPassword):
+        """Sets the user's password.
+
+        Arguments:
```
Comment (missgreenwood) [inline models/user.py]: docstring comment style

File: routes/login.py [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+from models.user import User, ActiveType
+from util.error import Error
+
+Login = Blueprint('Login', __name__)
+
+
+@Login.route('/login', methods=['POST'])
+def login():
+    """Performs a login for the given user credentials.
+    Parameters:
```
Comment (missgreenwood) [inline routes/login.py]: docstring comment style

File: routes/login.py [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+      String -- user: The user's name to login.
+      String -- password: The user's password in plaintext to login.
+    Returns:
+      200: If the login was successfull.
```
Comment (missgreenwood) [inline routes/login.py]: successful

File: routes/login.py [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+      401: If either user name was not found or password does not match.
+      404: If the useraccount is not active.
+    """
+    j = request.get_json(force=True)
+    user_requested = j['user']
+    password_requested = j['password']
+
+    targetUsers = User.query.filter_by(name=user_requested).all()
```
Comment (missgreenwood) [inline routes/login.py]: I don't really understand this. Why should there be more than one target user which matches the user name, given that the user name is unique as stated in models/user.py?

File: routes/set_password.py [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+                                  'There is no activation pending for this user.')
+    PASSWORD_TOO_SHOORT = Error('Password too short',
+                                'Provide a longer password and try again')
+
+
+@SetPassword.route('/set-password', methods=['POST'])
+def setPassword():
+    """ Set the password for an user who's activation is still pending.
+
+    Parameters:
```
Comment (missgreenwood) [inline routes/set_password.py]: Cf. what I wrote on strings and docstring comment format.

File: routes/set_password.py [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+      String -- user: The target user's name.
+      String -- token: The target user's token.
+      String -- password: The target user's password in plaintext to set.
+    Returns:
+        200: If the account has been activated successfully.
+        400: If activation_token does not match or user account is not in state pending.
+        404: If User is not found.
```
Comment (missgreenwood) [inline routes/set_password.py]: user

File: routes/set_password.py [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+
+
+@SetPassword.route('/set-password', methods=['POST'])
+def setPassword():
+    """ Set the password for an user who's activation is still pending.
+
+    Parameters:
+      String -- user: The target user's name.
+      String -- token: The target user's token.
+      String -- password: The target user's password in plaintext to set.
```
Comment (missgreenwood) [inline routes/set_password.py]: Do we really want to transfer the password in plain text? If it is encrypted at a different place, where?

File: routes/set_password.py [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+        200: If the account has been activated successfully.
+        400: If activation_token does not match or user account is not in state pending.
+        404: If User is not found.
+    """
+    j = request.get_json(force=True)
+    user_requested = j['user']
+    token = j['token']
+    password_requested = j['password']
+
+    targetUsers = User.query.filter_by(_name=user_requested).all()
```
Comment (missgreenwood) [inline routes/set_password.py]: Cf. what I wrote about unique user names.

File: routes/set_password.py [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+
+    if not any(targetUsers):
+        return Errors.UNKNOWN_USER.to_json(), status.HTTP_404_NOT_FOUND
+
+    targetUser = targetUsers[0]
+
+    if targetUser.get_activation_token() != token:
+        return (Errors.WRONG_TOKEN.to_json(),
```
Comment (missgreenwood) [inline routes/set_password.py]: Please use a unified syntax for tuples.

File: routes/set_password.py (hunk 30 above)
Comment (clemensmanert) [reply routes/set_password.py]: Plaintext just refers to the parameter.

**Pull request merged**